cpp/arcticdb/storage/azure/azure_storage-inl.hpp

```diff
@@ -0,0 +1,340 @@
+                KeyBucketizer&& bucketizer,
+                ReadKeyOpts opts,
+                const Azure::Storage::Blobs::DownloadBlobToOptions& download_option,
+                unsigned int request_timeout) {
+    ARCTICDB_SAMPLE(AzureStorageRead, 0)
+    auto fmt_db = [](auto&& k) { return variant_key_type(k); };
+    std::vector<VariantKey> failed_reads;
+
+    (fg::from(ks.as_range()) | fg::move | fg::groupBy(fmt_db)).foreach(
+        [&container_client, &root_folder, b=std::move(bucketizer), &visitor, &failed_reads,
+         opts=opts, &download_option, &request_timeout] (auto&& group) {
+        for (auto& k : group.values()) {
+            auto key_type_folder = object_store_utils::key_type_folder(root_folder, variant_key_type(k));
+            auto blob_name = object_store_utils::object_path(b.bucketize(key_type_folder, k), k);
+            try{
+                auto blob_client = container_client.GetBlockBlobClient(blob_name);
+                auto properties = blob_client.GetProperties(Azure::Storage::Blobs::GetBlobPropertiesOptions{}, get_context(request_timeout)).Value;
+                std::shared_ptr<Buffer> buffer = std::make_shared<Buffer>(properties.BlobSize);
+                blob_client.DownloadTo(buffer->preamble(), buffer->available(), download_option, get_context(request_timeout));
+                ARCTICDB_SUBSAMPLE(AzureStorageVisitSegment, 0)
+                visitor(k, Segment::from_buffer(std::move(buffer)));
+
+                ARCTICDB_DEBUG(log::storage(), "Read key {}: {}", variant_key_type(k), variant_key_view(k));
+            }
+            catch (const Azure::Core::RequestFailedException& e){
+                if (!opts.dont_warn_about_missing_key) {
+                    log::storage().warn("Failed to read azure segment with key '{}' {} {}: {}",
+                                            k,
+                                            blob_name,
+                                            static_cast<int>(e.StatusCode),
+                                            e.ReasonPhrase);
+                }
+                failed_reads.push_back(k);
+            }
+        }
+    });
+    if(!failed_reads.empty())
+        throw KeyNotFoundException(Composite<VariantKey>{std::move(failed_reads)});
+}
+
+template<class KeyBucketizer>
+void do_remove_impl(Composite<VariantKey>&& ks,
+                    const std::string& root_folder,
+                    Azure::Storage::Blobs::BlobContainerClient& container_client,
+                    KeyBucketizer&& bucketizer,
+                    bool connect_to_azurite,
+                    unsigned int request_timeout) {
+    ARCTICDB_SUBSAMPLE(AzureStorageDeleteBatch, 0)
+    auto fmt_db = [](auto&& k) { return variant_key_type(k); };
+    std::vector<VariantKey> failed_deletes;
+    if (connect_to_azurite){ //https://github.com/Azure/Azurite/issues/1822 due to an open issue of Azurite not filling subrequeset response after batch delete
```

Comment:
We can't really branch off an entire function for a test fixture library. It defeats the purpose of testing...

Reply:
Currently we can only test against Azurite (the enumerator)... I have raised this concern in the meeting with MS. I have test this function against a testing Azure storage account and it works fine.
In the near future I have to stick to Azurite for testing.
I understand that only code fully test should be released but it is a bit sad the code is not released due to the limit of enumerator. What is your thought?

Reply:
It has been amended. You may take a look again. More code is shared now.

Reply:
Hmmm the Azurite ticket is still opened and there is other person reporting the same issue, even after the fix.